compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1166,26 +1168,36 @@ private BAttachedFunction getNextFunc(BObjectType iteratorType) {
 
     public BType inferRecordFieldType(BRecordType recordType) {
         List<BField> fields = recordType.fields;
+        BUnionType unionType = BUnionType.create(null);
 
-        // If there are no fields in the record, return the rest field type as the inferred type.
-        if (fields.isEmpty()) {
-            return recordType.restFieldType;
+        if (recordType.restFieldType.tag != TypeTags.NONE) {
+            unionType.add(recordType.restFieldType);
         }
 
-        BType inferredType = fields.get(0).type; // If all the fields are the same, doesn't matter which one we pick
+        for (BField field : fields) {
+            if (!isAssignable(unionType, field.type) && !isAssignable(field.type, unionType)) {
```

Comment:
Should we extract out `isAssignable(unionType, field.type)` and `isAssignable(field.type, unionType)` to two `boolean` variables since we seem to be checking all of them twice, worst case?

Reply:
Should we also swap the checks?

L1183-L1185 first, L1187-L1190 second and if both fail we do this?

Reply:
fixed